### PR TITLE
Hdrp/refactor diffusion profile asset system

### PR DESCRIPTION
**Pull request by alelievr**
### Purpose of this PR
This PR improve the diffusion profile asset system to allow the distribution of diffusion profile assets (via asset store for example) without breaking your current diffusion profiles.
What does this changes ?
- Now there is only one diffusion profile per asset
- You can manage a list of used diffusion profiles per HD asset (default diffusion profiles)
- As a workaround to the limit of 16 diffusion profile on screen you now have a diffusion profile override component which allow to use specific diffusion profile inside volumes (or scene if the volume is global)
- Materials, shader graphs and VFX graph now references diffusion profiles asset directly (no more index)

---
### Testing status
**Katana Tests**: [Running](https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=HDRP%2FRefactorDiffusionProfileAssetSystem&automation-tools_branch=add-platform-filter&unity_branch=trunk#) 🤞 🤞 🤞

**Manual Tests**: local test, HDRP test project, VFX test project (not enough time to test on fontainebleau)

**Automated Tests**: no

---
### Overall Product Risks
**Technical Risk**:  High

**Halo Effect**: High

---
### Comments to reviewers
Adding reviewers to cover all the changes:
@sebastienlagarde for evrything
@Chman for volumes (and the fill profiles button)
@fredpointzero for slight API change in the migration API
@RSlysz for HDRP UI, volume UI and HDRP Wizard
@PaulDemeulenaere for VFX
@RemyUnity for QA
#### tips: huge changes, use the filter file to hide materials, meta, shadergraph, etc.

### CODE REVIEW

**Comment by sebastienlagarde** [comment]
Note: Several test are failing currently on HDRP (27)

File: com.unity.visualeffectgraph/Editor/Models/Contexts/Implementations/HDRP/VFXAbstractParticleHDRPLitOutput.cs [hunk 1]
```diff
@@ -51,8 +51,10 @@ public enum BaseColorMapMode
         [VFXSetting(VFXSettingAttribute.VisibleFlags.InInspector), SerializeField]
         protected bool onlyAmbientLighting = false;
 
-        [VFXSetting(VFXSettingAttribute.VisibleFlags.InInspector), SerializeField, Range(1, 15)]
-        protected uint diffusionProfile = 1;
+#if VFX_HAS_HDRP
```
Comment (sebastienlagarde) [inline com.unity.visualeffectgraph/Editor/Models/Contexts/Implementations/HDRP/VFXAbstractParticleHDRPLitOutput.cs]: isn't this dangerous to have szeialized field based on a define?

File: com.unity.render-pipelines.high-definition/Editor/Lighting/DiffusionProfileOverrideEditor.cs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                DiffusionProfileSettings defaultValue = (index < m_DefaultDiffusionProfiles.Length) ? m_DefaultDiffusionProfiles[index] : null;
+                EditorGUI.BeginDisabledGroup(true);
+                EditorGUI.ObjectField(rect, new GUIContent("Profile " + index), defaultValue, typeof(DiffusionProfileSettings), false);
+                EditorGUI.EndDisabledGroup();
+            }
+            else
+                EditorGUI.ObjectField(rect, element, new GUIContent("Profile " + index));
+        }
+
+        void FillProfileListWithScene()
```
[the remark this answers is not in the document]
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/DiffusionProfileOverrideEditor.cs]: Yes it works automagically :)

**Comment by PaulDemeulenaere** [comment]
Here a variant of VisualEffectGraph on Legacy : [VisualEffectGraph_Legacy.zip](https://github.com/Unity-Technologies/ScriptableRenderPipeline/files/2875468/VisualEffectGraph_Legacy.zip) <= To Unzip in TestProjects folder

HDRPLitOutput isn't expected to work but you can try to launch graphicTest and verify you don't have any compilation error.

File: com.unity.render-pipelines.high-definition/Editor/RenderPipeline/HDRenderPipelineUI.cs [hunk 3]
```diff
@@ -432,6 +435,11 @@ static void Drawer_SectionMaterialUnsorted(SerializedHDRenderPipelineAsset seria
             EditorGUILayout.PropertyField(serialized.renderPipelineSettings.lightLoopSettings.supportFabricConvolution, k_SupportFabricBSDFConvolutionContent);
         }
 
+        static void DrawDiffusionProfileElement(SerializedProperty element, Rect rect, int index)
+        {
+            EditorGUI.ObjectField(rect, element, new GUIContent("Profile " + index));
```
[the remark this answers is not in the document]
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/RenderPipeline/HDRenderPipelineUI.cs]: done, I used `TrTextContent`

File: com.unity.render-pipelines.high-definition/Editor/Material/Lit/LitUI.cs [hunk 4]
```diff
@@ -446,57 +450,44 @@ protected void ShaderSSSAndTransmissionInputGUI(Material material, int layerInde
...
                 using (new EditorGUILayout.HorizontalScope())
                 {
-                    EditorGUILayout.PrefixLabel(Styles.diffusionProfileText);
+                    // We can't cache these fields because of several edge cases like undo/redo or pressing escape in the object picker
+                    string guid = HDUtils.ConvertVector4ToGUID(diffusionProfileAsset[layerIndex].vectorValue);
+                    DiffusionProfileSettings diffusionProfile = AssetDatabase.LoadAssetAtPath<DiffusionProfileSettings>(AssetDatabase.GUIDToAssetPath(guid));
 
-                    using (new EditorGUILayout.HorizontalScope())
+                    // is it okay to do this every frame ?
+                    EditorGUI.BeginChangeCheck();
```
[the remark this answers is not in the document]
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Material/Lit/LitUI.cs]: done

File: com.unity.render-pipelines.high-definition/Editor/Lighting/DiffusionProfileOverrideEditor.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+        public override void OnEnable()
+        {
+            var o = new PropertyFetcher<DiffusionProfileOverride>(serializedObject);
+
+            m_Volume = (m_Inspector.target as Volume);
+            m_DiffusionProfiles = Unpack(o.Find(x => x.diffusionProfiles));
+            m_OverrideStates = Unpack(o.Find(x => x.overrideStates));
+            var hdAsset = GraphicsSettings.renderPipelineAsset as HDRenderPipelineAsset;
+            m_DefaultDiffusionProfiles = hdAsset.diffusionProfileSettingsList;
```
[the remark this answers is not in the document]
Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/DiffusionProfileOverrideEditor.cs]: fixed (code removed)

**Comment by alelievr** [comment]
Done the VFX legacy test it compiles so the VFX_HDRP define is working

**Comment by RemyUnity** [comment]
The menu item to create a diffusion profile is still "Rendering/**Diffusion Profile List**" and the new asset default name is "**New Diffusion Profile List**" when it should not be a list anymore.

File: com.unity.render-pipelines.high-definition/Runtime/Material/Lit/Lit.shader [hunk 6]
```diff
@@ -217,6 +216,9 @@ Shader "HDRP/Lit"
         [ToggleUI] _SupportDecals("Support Decals", Float) = 1.0
         [ToggleUI] _ReceivesSSR("Receives SSR", Float) = 1.0
 
+        [HideInInspector] _DiffusionProfile("Obsolete, kept for migration purpose", Int) = 0
+        [HideInInspector] _DiffusionProfileAsset("Dissuion Profile Asset", Vector) = (0, 0, 0, 0)
```
Comment (fabien-unity) [inline com.unity.render-pipelines.high-definition/Runtime/Material/Lit/Lit.shader]: typo 'Dissuion'. it's hidden, so maybe not important

Comment (alelievr) [reply com.unity.render-pipelines.high-definition/Runtime/Material/Lit/Lit.shader]: fixed

**Pull request merged**